docs/DesignDocs/system_ocall_opt_in.md

```diff
@@ -0,0 +1,120 @@
+if an enclave with edl file `sample.edl` wants to use the function `oe_log_ocall()` they can do
+one of the following:
+
+*Note: These examples assume that `${install_path}/include` is passed to edger8r as an edl search path*
+
+1. Import just the required OCall
+
+```
+// sample.edl
+enclave
+{
+    from "openenclave/edl/tee.edl" import oe_log_ocall;
+
+    trusted
+    {
+        public void my_sample_ecall();
+    };
+
+    untrusted
+    {
+        public void my_sample_ocall();
+    };
+}
+```
+
+2. Import all OCalls from `tee.edl`
+
+```
+// sample.edl
+enclave
+{
+    from "openenclave/edl/tee.edl" import *;
+
+    trusted
+    {
+        public void my_sample_ecall();
+    };
+
+    untrusted
+    {
+        public void my_sample_ocall();
+    };
+}
+```
+
+In example (1), linking would fail if the enclave made a call to `oe_realloc_ocall()`
+(also part of tee.edl) unless the developer modified the import line to
+`from "openenclave/edl/tee.edl" import oe_log_ocall, oe_realloc_ocall;`. In this scenario,
+example (2) will work without modification.
+
+## EDL local structures
+
+Under this design, system edl files will prefer using local structures; that is, structures
+that are defined in edl. This is distinct from foreign structures, which are defined in
+C/C++ header files and included in the edl file.
+
+In the case that a foreign structure must be used, system edl files may only include headers
+that are installed with the SDK.
+
+Local EDL structures provide additional security benefits detailed [here](
+https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/full_edl_serialization.md#motivation)
+so foreign structures should be avoided whenever possible.
+
+## Developer opt-out
+
+Because system OCalls are generally called by the SDK implementation, developers
+may not have direct control over which OCalls will be called by their application.
+A developer can refuse to import a function in EDL which will cause a linker error
+if it is used, but they should also be able to take action to resolve this failure.
+
+To this end, any OE SDK feature which makes an OCall on behalf of an enclave
+should provide a mechanism to disable the OCall at compile-time and fail
+gracefully.
+
+If this is not possible, the feature must be able to be disabled at the time
+of OE SDK compilation.
+
+## Exceptions to edl import and opt-in/out mechanisms
+
+A small number of OCalls in OE SDK are not implemented in edl and cannot be for
+technical reasons. These OCalls will always be included and the developer cannot
+opt out of them because they support core SDK functionality.
+
+These calls are:
+* `OE_OCALL_MALLOC`
+* `OE_OCALL_FREE`
+* `OE_OCALL_THREAD_WAIT`
+* `OE_OCALL_THREAD_WAKE`
+* `OE_OCALL_WAKE_HOST_WORKER`
+
+## List of system OCalls
+
+All system OCalls (other than the previously mentioned exceptions) are found
+in one of the following edl files
+
+* `common/tee.edl`
+* `common/syscall.edl`
```

Comment:
This coarse granularity means you might need a long list of function-level imports/exports if you want the granularity.
Why not split these into finer grained EDL files, e.g., separate out sockets, stdio, etc. (This is what we did in new_platforms)

Reply:
Agreed, my goal is to split them out, but I haven't had a chance to do the work yet and I wanted to settle on a design first.

Reply:
Opened #2585 to track this